### PR TITLE
kube-apiserver: fix TLS handshake errors on Packet

### PR DESCRIPTION
**Pull request by invidian**
See commit messages for details.

Refs #278

### CODE REVIEW

**Comment by invidian** [comment]
Note: the documentation for that will be added as part of #295, as currently we don't have a good place to document it.

File: assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml [hunk 1]
```diff
@@ -239,8 +239,7 @@ spec:
 
           {{ end }}
           backend kube-apiserver
-            option httpchk GET /healthz HTTP/1.1\r\nHost:\ kube-apiserver\r\nAuthorization:\ Bearer\ {{ template "token" . }}
-            server 1 $ADDRESS:{{ template "port" . }} verify none check check-ssl
+            server 1 $ADDRESS:{{ template "port" . }}
```
Comment (iaguis) [inline assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml:242]: should we squash this with the previous commit? can we explain it a bit more (for example, why don't we have the `check` keyword now?)?

File: assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml [hunk 2]
```diff
@@ -221,9 +239,11 @@ spec:
 
           {{ end }}
           backend kube-apiserver
-            server 1 $ADDRESS:{{ template "port" . }} check
+            option httpchk GET /healthz HTTP/1.1\r\nHost:\ kube-apiserver\r\nAuthorization:\ Bearer\ {{ template "token" . }}
+            server 1 $ADDRESS:{{ template "port" . }} verify none check check-ssl
           EOF
           cat /run/kube-apiserver/haproxy.cfg
+          echo "{{ template "token" . }},nobody,nobody" > /run/kube-apiserver/token-auth-file
```
Comment (iaguis) [inline assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml:245]: I was trying to find info on this "nobody" user but couldn't find anything. Do you have a reference here? Or is it just a non-existent user and that means they don't have any privileges?

File: assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml [hunk 3]
```diff
@@ -14,6 +14,12 @@
 6443
 {{- end -}}
 {{ end }}
+{{- define "token" -}}
+{{ include (print $.Template.BasePath "/kube-apiserver-secret.yaml") . | sha256sum }}
```
Comment (iaguis) [inline assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml:18]: I assume this is the rendered file, right?

File: assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml (hunk 1 above)
Comment (invidian) [reply assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml:242]: The explanation is in the commit:

>As we always have only one backend, it doesn't make sense to perform
those checks.

Simply, HAProxy with checks can remove unhealthy backend from the list of backends, so the requests does not fail. In our case, we only ever have one backend, so we don't need to have health checks, as HAProxy won't return any cached or more specific errors either.

What I think we should have is the liveness probes for both containers, so they are restarted if they stop responding. But I think this is the topic for another issue.

I can actually swap the order of the commits, so we don't add it and then remove it back.

File: assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml (hunk 2 above)
Comment (invidian) [reply assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml:245]: It's just an arbitrary user name, which do not have any RBAC associated with. I wouldn't say it's "non-existent", as K8s users are not really stateful. If it has valid API token/cert, it appears in the system :smile: 

If you have better suggestions, I'm happy to adopt.

File: assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml (hunk 3 above)
Comment (invidian) [reply assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml:18]: Yes, this is already a template with values included. The pattern is from https://helm.sh/docs/howto/charts_tips_and_tricks/ "https://helm.sh/docs/howto/charts_tips_and_tricks/" section.

File: assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml (hunk 2 above)
Comment (iaguis) [reply assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml:245]: that's fine then

File: assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml (hunk 3 above)
Comment (iaguis) [reply assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml:18]: cool

File: assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml (hunk 1 above)
Comment (iaguis) [reply assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml:242]: ok, a verbose explanation like this comment would be better than the on in the commit :)

File: assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml (hunk 2 above)
Comment (invidian) [reply assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml:245]: I'll add an inline comment about it for the time being.

File: assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml (hunk 3 above)
Comment (invidian) [reply assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml:18]: Ditto inline comment.

File: assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml (hunk 1 above)
Comment (invidian) [reply assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml:245]: Added.

**Comment by iaguis** [comment]
CI is broken.

**Pull request merged**